New JS interop docs (#4578)
dart-lang/site-www: src/interop/js-interop/index.md

---
title: JavaScript interoperability
short-title: JS interop
description: Integrate JavaScript code into your Dart web app.
---

The [Dart web platform](/overview#web-platform) supports communication with
JavaScript apps and libraries, and browser DOM APIs, using the `js_interop`

> package, also known as `dart:js_interop`.

Reviewer: ```suggestion
library, also known as `dart:js_interop`.
```

It's not a package! It's a (`dart:`) library!

Web developers can benefit from using external JS libraries in their Dart code, without
having to rewrite anything in Dart.

## Static interop

The principal model of JS interop in Dart is **static interop**.
Static interop means interop that requires staticly typing external members,
rather than allowing dynamic invocations.
This enables [features](#features) like better performance, type soundness, and more.

> Performant static interop model is made possible by **[inline classes][]**.
> Inline classes are a special class type that wrap an existing type into a new static type,
> without the overhead of a traditional class, enabling zero cost wrapping.

Reviewer 2: Perhaps:
```suggestion
Performant static interop model is made possible by **[extension types][]**.
Extension types are a special class type that wrap an existing type into a new static type,
without the overhead of a traditional class, enabling zero cost wrapping.
```

### Why static interop?

Static interop represents the desire to decouple Dart's "core" from the platform
where it's embedded
(*maybe definition of "embedded" or examples here, like the browser, Flutter, etc.?*).

Our [previous iterations of JS interop][] provided the capabillity

> to access the embedder, but were not primed to handle that decoupling.

Reviewer: I don't understand what "embedder" is here.

> Too much of the process was entwined into the Dart side,

Reviewer: I'm worried about ESL folks reading this. It's tricky to parse...

like writing all the bindings to the browser
(*not sure if I understood that note correctly*). 
They also had limitations around using the DOM,
typing (which is a cornerstone of Dart),
and expanding to new interop types
(*this is me trying to refer to wasm without actually saying, idk if "interop types" is the right term*). 

Static interop addresses all of the shortcomings of Dart's previous JS interop solutions.
Check out the [Features](#features) section for a summary of improvements. 

[previous iterations of JS interop]: /interop/js-interop/past-js-interop

## Usage

The following example implements inline class-based, static, JS interop in Dart,
using its key members and syntaxes:

```dart
@JS()
library;
// library names aren't cool anymore...

import 'dart:js_interop';

> inline class SomeThing {

Reviewer: ```suggestion
extension type SomeThing {
```

Not quite what we want, but it's closer!
